File: python/stack_01.py
# 2020年 04月 16日 星期四 14:57:00 CST
# 栈的python实现，基于列表
MAXSTACK = 100
global stack
stack = [None] * MAXSTACK
top = -1

def isEmpty():
    if top == -1:
        return True
    else:
        return False

def push(data):
    global top
    global MAXSTACK
    global stack
    if top >= MAXSTACK:
        print("堆栈已满，无法加入")
    else:
        top += 1
        stack[top] = data

def pop():
    global top
    global stack
    if isEmpty():
        print("堆栈是空的")
    else:
        print("弹出的元素是：%d" % stack[top])
        top -= 1

i = 2
count = 0
while True:
    i = int(input("要压入堆栈，请输1；要弹出请输0；结束输入-1"))
    if i == -1:
        break
    elif i == 1:
        val = int(input('请输入元素值：'))
        push(val)
    elif i == 0:
        pop()

print("============")
if top < 0:
    print("堆栈已空")
else:
    i = top
    while i >= 0:
        print("堆栈弹出顺序为：%d" % stack[i])
        count += 1
        i -= 1
print("============")

 

File: c/stack_01.c
/* 
 *2020年 04月 16日 星期四 15:17:22 CST
 *基于数组的堆栈实现
*/
#include<stdio.h>
#include<time.h>
#include<stdlib.h>

#define max_length 100
typedef struct Stack {
    int data[max_length];
    int top;
}stack;

int is_empty(stack const * const s)
{
    return s->top == -1 ;
}

void push(stack *const s, int data)
{
    if (s->top >= (max_length - 1)) {
        printf("stack is full!\n");
        return;
    }
    else {
        s->top++;
        s->data[s->top] = data;
    }
}

void pop(stack *s)
{
    if (is_empty(s)) {
        printf("stack is empty!\n");
        return;
    }
    else {
        printf("the top element is %d\n", s->data[s->top--]);
    }
}

int main(void)
{
    stack s;
    s.top = -1;
    srand((int)time(0));
    int flag = 1;
    int value = 0;
    while (flag != -1) {
        value = rand() % 30;
        printf("rand = %d\n", value);
        printf("push: input 1; pop: input 0; exit: input -1:   ");
        scanf("%d", &flag);
        if (flag == 1)
            push(&s, value);
        else if (flag == 0)
            pop(&s);
    }
    printf("============\n");
    for (int i = s.top; i >= 0; i--)
        pop(&s);
}